src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Unix.cs

```diff
@@ -770,144 +760,162 @@ private static void WriteRandomMask(byte[] buffer, int offset)
             {
-                while (true) // in case we get control frames that should be ignored from the user's perspective
+                CancellationTokenRegistration registration = cancellationToken.Register(s => ((ManagedClientWebSocket)s).Abort(), this);
+                try
                 {
-                    // Get the last received header.  If its payload length is non-zero, that means we previously
-                    // received the header but were only able to read a part of the fragment, so we should skip
-                    // reading another header and just proceed to use that same header and read more data associated
-                    // with it.  If instead its payload length is zero, then we've completed the processing of
-                    // thta message, and we should read the next header.
-                    MessageHeader header = _lastReceiveHeader;
-                    if (header.PayloadLength == 0)
-                    {
-                        MessageHeader? headerOpt = await ReadMessageHeaderAsync(cancellationToken).ConfigureAwait(false);
-                        if (headerOpt == null)
-                        {
-                            // The connection closed; nothing more to read.
-                            return new WebSocketReceiveResult(0, WebSocketMessageType.Text, true);
-                        }
-                        header = headerOpt.GetValueOrDefault();
-                    }
-
-                    // If the header represents a ping or a pong, handle it.
-                    if (header.Opcode == MessageOpcode.Ping || header.Opcode == MessageOpcode.Pong)
+                    while (true) // in case we get control frames that should be ignored from the user's perspective
                     {
-                        // Consume any (optional) payload associated with the ping/pong
-                        if (header.PayloadLength > 0 && _receiveBufferCount < header.PayloadLength)
+                        // Get the last received header.  If its payload length is non-zero, that means we previously
+                        // received the header but were only able to read a part of the fragment, so we should skip
+                        // reading another header and just proceed to use that same header and read more data associated
+                        // with it.  If instead its payload length is zero, then we've completed the processing of
+                        // thta message, and we should read the next header.
+                        MessageHeader header = _lastReceiveHeader;
+                        if (header.PayloadLength == 0)
                         {
-                            await EnsureBufferContainsAsync((int)header.PayloadLength, cancellationToken).ConfigureAwait(false);
-                        }
+                            if (_receiveBufferCount < MaxMessageHeaderLength && // fast check to see if we have enough data for any possible header
```

Comment:
> O...kay... I think I understand this if: If we have received MaxMessageHeaderLength bytes than we know that we have a completed header, so don't create the task.

Right.

The header is the one outlined here:
https://github.com/dotnet/corefx/blob/master/src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Unix.cs#L689
It's either 6 bytes, 8 bytes, or 14 bytes, depending on the length of the payload:
- Payloads less than or equal to 125 bytes have a 6 byte header
- Payloads between 126 and ushort.MaxValue bytes have an 8 byte header, the extra two bytes being used to store the 16-bit length
- Payloads longer than ushort.MaxValue have a 14 byte header, the extra eight bytes being used to store the 64-bit length
So the max header length is 14 bytes... if we have 14 bytes already in our buffer, there's no need to read anything yet, as we have enough to parse.

> Is there a test that hits the boundary condition here?

I expect we'd have to have a custom localhost server we could connect to and use it to control how much data was delivered and when.  We don't currently have that.